src/Microsoft.ML.TimeSeries/RootCauseAnalyzer.cs

```diff
@@ -0,0 +1,738 @@
+                    if (anomalyPoint != null && rootCausePoint != null)
+                    {
+                        Tuple<double, double> scores = GetSupriseAndExplainaryScore(rootCausePoint, anomalyPoint);
+                        scoreList.Add(new RootCauseScore(scores.Item1, scores.Item2));
+                        item.Direction = GetRootCauseDirection(rootCausePoint);
+                    }
+                }
+
+                //get final score
+                for (int i = 0; i < scoreList.Count; i++)
+                {
+                    dst.Items[i].Score = GetFinalScore(scoreList[i].Surprise, Math.Abs(scoreList[i].ExplainaryScore), beta);
+                }
+            }
+            else if (dst.Items.Count == 1)
+            {
+                Point rootCausePoint = GetPointByDimension(dimPointMapping, dst.Items[0].Dimension, pointTree, aggType, aggSymbol);
+                if (anomalyPoint != null && rootCausePoint != null)
+                {
+                    Tuple<double, double> scores = GetSupriseAndExplainaryScore(rootCausePoint, anomalyPoint);
+                    dst.Items[0].Score = GetFinalScore(scores.Item1, scores.Item2, beta);
+                    dst.Items[0].Direction = GetRootCauseDirection(rootCausePoint);
+                }
+            }
+        }
+
+        private Point GetPointByDimension(Dictionary<string, Point> dimPointMapping, Dictionary<string, Object> dimension, PointTree pointTree, AggregateType aggType, string aggSymbol)
+        {
+            if (dimPointMapping.ContainsKey(GetDicCode(dimension)))
+            {
+                return dimPointMapping[GetDicCode(dimension)];
+            }
+
+            int count = 0;
+            Point p = new Point(dimension);
+            DimensionInfo dimensionInfo = SeperateDimension(dimension, aggSymbol);
+            Dictionary<string, Object> subDim = GetSubDim(dimension, dimensionInfo.DetailDims);
+
+            foreach (Point leave in pointTree.Leaves)
+            {
+                if (ContainsAll(leave.Dimension, subDim))
+                {
+                    count++;
+
+                    p.Value = +leave.Value;
+                    p.ExpectedValue = +leave.ExpectedValue;
+                    p.Delta = +leave.Delta;
+                }
+
+            }
+            if (aggType.Equals(AggregateType.Avg))
+            {
+                p.Value = p.Value / count;
+                p.ExpectedValue = p.ExpectedValue / count;
+                p.Delta = p.Delta / count;
+            }
+
+            if (count > 0)
+            {
+                return p;
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        private static string GetDicCode(Dictionary<string, Object> dic)
+        {
+            return string.Join(";", dic.Select(x => x.Key + "=" + (string)x.Value).ToArray());
```

Comment:
This seems inefficient/unnecessary? Are you not able to use just use "Dictionary<string, Object>" as the dictionary key (where you use GetDicCode)?

Reply:
> This seems inefficient/unnecessary? Are you not able to use just use "Dictionary<string, Object>" as the dictionary key (where you use GetDicCode)?

I have tried to use the Dictionary while overriding the equals function, but it seems not work well, so I convert this to a string, so we can compare it.